src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1668,26 +1668,50 @@ private DiagnosticInfo GetBadEventUsageDiagnosticInfo(EventSymbol eventSymbol)
+            var receiver = propertyAccess.ReceiverOpt;
+            if (!propertySymbol.IsDefinition && propertySymbol.ContainingType.Equals(propertySymbol.ContainingType.OriginalDefinition, TypeCompareKind.IgnoreNullableModifiersForReferenceTypes))
+            {
+                propertySymbol = propertySymbol.OriginalDefinition;
+            }
+
+            return propertySymbol is SourcePropertySymbolBase sourceProperty &&
+                sourceProperty.GetMethod is SourcePropertyAccessorSymbol { IsEquivalentToBackingFieldAccess: true } &&
+                TypeSymbol.Equals(sourceProperty.ContainingType, fromMember.ContainingType, TypeCompareKind.ConsiderEverything2) &&
+                    IsConstructorOrField(fromMember, isStatic: sourceProperty.IsStatic) &&
+                    (sourceProperty.IsStatic || receiver.Kind == BoundKind.ThisReference);
+        }
+
+        internal static bool IsPropertyAssignedThroughBackingField(BoundPropertyAccess propertyAccess, Symbol fromMember)
+        {
+            return IsPropertyAssignedThroughBackingField(propertyAccess.ReceiverOpt, propertyAccess.PropertySymbol, fromMember);
         }
 
-        private static bool AccessingAutoPropertyFromConstructor(BoundExpression receiver, PropertySymbol propertySymbol, Symbol fromMember)
+        /// <summary>
+        /// Determines whether a property is assigned through the backing field within a symbol <paramref name="fromMember"/>
+        /// </summary>
+        /// <remarks>
+        /// A property is assigned through backing field within a constructor or a field assignment.
+        /// </remarks>
+        private static bool IsPropertyAssignedThroughBackingField(BoundExpression receiver, PropertySymbol propertySymbol, Symbol fromMember)
         {
             if (!propertySymbol.IsDefinition && propertySymbol.ContainingType.Equals(propertySymbol.ContainingType.OriginalDefinition, TypeCompareKind.IgnoreNullableModifiersForReferenceTypes))
             {
                 propertySymbol = propertySymbol.OriginalDefinition;
             }
 
-            var sourceProperty = propertySymbol as SourcePropertySymbolBase;
-            var propertyIsStatic = propertySymbol.IsStatic;
+            // PROTOTYPE(semi-auto-props): TODO: Support assigning semi auto prop from constructors.
 
-            return (object)sourceProperty != null &&
-                    sourceProperty.IsAutoPropertyWithGetAccessor &&
+            return propertySymbol is SourcePropertySymbolBase sourceProperty &&
+                    sourceProperty.GetMethod is not null &&
+                    // To be assigned through backing field, either SetMethod is null, or it's equivalent to backing field write
+                    // PROTOTYPE(semi-auto-props): TODO: Do we need to use `GetOwnOrInheritedSetMethod` instead of `SetMethod`? Add tests.
+                    sourceProperty.SetMethod is not SourcePropertyAccessorSymbol { IsEquivalentToBackingFieldAccess: false } &&
```

Comment:
>sourceProperty.SetMethod is not SourcePropertyAccessorSymbol { IsEquivalentToBackingFieldAccess: false } [](http://example.com/codeflow?start=20&length=104)

This condition doesn't quite match the stated intent: "either SetMethod is null, or it's equivalent to backing field write"

Reply:
> > sourceProperty.SetMethod is not SourcePropertyAccessorSymbol { IsEquivalentToBackingFieldAccess: false }
> 
> This condition doesn't quite match the stated intent: "either SetMethod is null, or it's equivalent to backing field write"

I think it does. If null we will return true. If equivalent to backing field access, we return true as well.

Reply:
> I think it does. If null we will return true. If equivalent to backing field access, we return true as well.

Can it return true under other conditions?

Reply:
@AlekseyTs Do you mean we have a type other than  `SourcePropertyAccessorSymbol`? I'll validate if this can happen.

Reply:
> Do you mean we have a type other than  ```SourcePropertyAccessorSymbol```?

Yes, for example.

> I'll validate if this can happen.

Not sure if you have some assert in mind as a validation. instead, I suggest rewriting the condition to implement the intent as stated and not making assumptions what other types could be used to represent an accessor symbol.